components/director/internal/runtimemapping/token_verifier.go

```diff
@@ -0,0 +1,263 @@
+package runtimemapping
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"path"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/dgrijalva/jwt-go"
+	"github.com/lestrrat-go/jwx/jwk"
+	"github.com/pkg/errors"
+)
+
+const (
+	bearerPrefix = "Bearer "
+
+	claimsIssuerKey = "iss"
+	claimsNameKey   = "name"
+	claimsGroupsKey = "groups"
+	claimsEmailKey  = "email"
+	jwksKeyIDKey    = "kid"
+	jwksURIKey      = "jwks_uri"
+
+	openIDConfigSubpath = ".well-known/openid-configuration"
+)
+
+type KeyGetter interface {
+	GetKey(token *jwt.Token) (interface{}, error)
+}
+
+type tokenVerifier struct {
+	logger Logger
+	keys   KeyGetter
+}
+
+func NewTokenVerifier(logger Logger, keys KeyGetter) *tokenVerifier {
+	return &tokenVerifier{
+		logger: logger,
+		keys:   keys,
+	}
+}
+
+func (tv *tokenVerifier) Verify(token string) (*jwt.MapClaims, error) {
+	if token == "" {
+		return nil, errors.New("token cannot be empty")
+	}
+
+	claims, err := tv.verifyToken(token)
+	if err != nil {
+		return nil, errors.Wrap(err, "while veryfing the token")
+	}
+
+	return claims, nil
+}
+
+func (tv *tokenVerifier) verifyToken(tokenStr string) (*jwt.MapClaims, error) {
+	tokenStr = strings.TrimPrefix(tokenStr, bearerPrefix)
+	claims := new(jwt.MapClaims)
+
+	_, err := new(jwt.Parser).ParseWithClaims(tokenStr, claims, tv.keys.GetKey)
+	if err != nil {
+		return nil, errors.Wrap(err, "while parsing the token with claims")
+	}
+
+	return claims, nil
+}
+
+type jwkCacheEntry struct {
+	key      interface{}
+	cachedAt time.Time
+}
+
+func (e jwkCacheEntry) IsExpired() bool {
+	return time.Now().After(e.cachedAt.Add(5 * time.Minute))
+}
+
+type jwksCache struct {
+	logger Logger
+	fetch  KeyGetter
+	cache  map[string]jwkCacheEntry
+	flag   sync.Mutex
+}
+
+func NewJWKsCache(logger Logger, fetch KeyGetter) *jwksCache {
+	return &jwksCache{
+		logger: logger,
+		fetch:  fetch,
+		cache:  make(map[string]jwkCacheEntry),
+	}
+}
+
+func (c *jwksCache) GetKey(token *jwt.Token) (interface{}, error) {
+	if token == nil {
+		return nil, errors.New("token cannot be nil")
+	}
+
+	keyID, err := getKeyID(*token)
+	if err != nil {
+		return nil, errors.Wrap(err, "while getting the key ID")
+	}
+
+	c.flag.Lock()
```

Comment:
maybe `sync.Map` will be easier to use.

Reply:
Well, due to sync.Map logic I prefer map with static typing than sync.Map where I need to cast the items to be able to deremin whether the item expired.